File: packages/py/tracing/src/uselemma_tracing/release.py
from __future__ import annotations

from typing import Any

RELEASE_MAX_LENGTH = 200

_CONTROL_CHARS = ("\n", "\t", "\r")


def normalize_release(value: Any) -> str | None:
    """Trim; drop empty; reject newlines/tabs/CR; cap at 200 chars.

    Invalid input is unreleased (``None``). Does not guess a git SHA.
    """
    if not isinstance(value, str):
        return None
    trimmed = value.strip()
    if not trimmed:
        return None
    if len(trimmed) > RELEASE_MAX_LENGTH:
        return None
    if any(char in trimmed for char in _CONTROL_CHARS):
        return None
    return trimmed

File: packages/ts/tracing/src/release.ts
export const RELEASE_MAX_LENGTH = 200;

const CONTROL_CHARS = /[\n\t\r]/;

/**
 * Trim; drop empty; reject newlines/tabs/CR; cap at 200 chars.
 * Invalid input is unreleased (`undefined`). Does not guess a git SHA.
 */
export function normalizeRelease(value: unknown): string | undefined {
  if (typeof value !== "string") return undefined;
  const trimmed = value.trim();
  if (trimmed.length === 0) return undefined;
  if (trimmed.length > RELEASE_MAX_LENGTH) return undefined;
  if (CONTROL_CHARS.test(trimmed)) return undefined;
  return trimmed;
}
